Do not reset connection immediately if grpc code is `Canceled` or `DeadlineExceeded`

congqixia opened the pull request:
We found lots of connection reset & canceled due to recent retry change
Current implementation resets connection no matter what the error code is
To sync behavior to previous retry, skip reset connection only if cancel error happens too much.

Also adds a config item for minResetInterval for grpc reset connection
/kind improvement

SimFG commented:
/lgtm

congqixia commented:
/run-cpu-e2e

congqixia commented:
/run-cpu-e2e

xiaofan-luan commented on internal/util/grpcclient/client.go:
```diff
@@ -328,12 +341,16 @@ func (c *ClientBase[T]) call(ctx context.Context, caller func(client T) (any, er
 		ret, callErr = caller(client)
 		if callErr != nil {
-			if funcutil.IsGrpcErr(callErr) ||
+			if funcutil.IsGrpcErr(callErr, codes.Unavailable, codes.Aborted, codes.FailedPrecondition) ||
```
I don't like the idea to list all the retriable errors.
But you can list errors you don't want to reset connection

xiaofan-luan commented on internal/util/grpcclient/client.go:
```diff
@@ -328,12 +341,16 @@ func (c *ClientBase[T]) call(ctx context.Context, caller func(client T) (any, er
+			}
 			if !funcutil.CheckCtxValid(ctx) {
 				if c.sess != nil {
 					sessions, _, getSessionErr := c.sess.GetSessions(c.GetRole())
```
this is might also be too frequent. 

We need to watch session in each querynode. and decide whether reset connection by if session exist

xiaofan-luan commented on internal/util/grpcclient/client.go:
```diff
@@ -328,12 +341,16 @@ func (c *ClientBase[T]) call(ctx context.Context, caller func(client T) (any, er
 		if callErr != nil {
-			if funcutil.IsGrpcErr(callErr) ||
+			if funcutil.IsGrpcErr(callErr, codes.Unavailable, codes.Aborted, codes.FailedPrecondition) ||
 				IsCrossClusterRoutingErr(callErr) || IsServerIDMismatchErr(callErr) {
```
are we pretty sure IsCrossClusterRoutingErr and IsServerIDMismatchErr throws with code 
codes.Unavailable, codes.Aborted, codes.FailedPrecondition?

congqixia replied:
This condition lists all the case we might wanna reset the connection here

congqixia committed:
Do not reset connection immediately if grpc code is `Canceled` or `DeadlineExceeded`

We found lots of connection reset & canceled due to recent retry change
Current implementation resets connection no matter what the error code is
To sync behavior to previous retry, skip reset connection only if cancel error happens too much.

Also adds a config item for minResetInterval for grpc reset connection

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

xiaofan-luan commented:
/lgtm
/approve

congqixia commented:
/run-cpu-e2e

congqixia commented:
last e2e failed due to compaction timeout

weiliu1031 commented:
/lgtm
